File: 2019/2019W42/preppindataw42.R
library(dplyr)
library(tidyr)
library(readxl)
library(RJSONIO)

checklocation = function(city_name,country_name) {
  city_name_fixed = gsub(' ', '%20', city_name)
  country_name_fixed = gsub(' ', '%20', country_name)
  url = paste0(
    "http://nominatim.openstreetmap.org/search?city=",
    city_name_fixed,
    "&country=",
    country_name_fixed,
    "&limit=1&format=json")
  resOSM = fromJSON(url)
  if(length(resOSM) > 0) {
    return(1)
  } else {
    return(0) 
  }
}

Sys.setlocale("LC_ALL", "English")

final <- read_excel(path = "E:/PD Wk 42 Input.xlsx", 
                   sheet = "Sheet1") %>%
  separate(.,'Location',c('City','Country'),sep=',\\s*') %>%
  separate(.,'Store Potential Sales',c('Store Potential Sales','Currency'),sep='\\s+',convert=TRUE) %>%
  separate(.,'Store Cost',c('Store Cost','Currency'),sep='\\s+',convert=TRUE) %>%
  rename('Zip Code' = 'Potential Store Location') %>%
  filter(`Store Potential Sales` > `Store Cost`) %>%
  rowwise() %>%
  mutate('Check' = checklocation(`City`,`Country`)) %>%
  ungroup() %>%
  filter(`Check`==1) %>%
  merge(.,read_excel(path = "E:/PD Wk 42 Input.xlsx", 
                     sheet = "Currency Conversion"),
        by = 'Currency') %>%
  group_by(`Zip Code`) %>%
  filter(`Store Potential Sales`== max(`Store Potential Sales`)) %>%
  select(c('City', 'Country', 'Zip Code', 'Store Potential Sales', 'Store Cost',
           'Currency', 'Value in USD'))

View(final)

File: 2019/2019W42/PreppindataW42.py
import pandas as pd
import geonamescache

gc = geonamescache.GeonamesCache()
countries = gc.get_countries_by_names()

xls = pd.ExcelFile("E:/PD Wk 42 Input.xlsx")
instance = pd.read_excel(xls,'Sheet1')
currency = pd.read_excel(xls,'Currency Conversion')

temp = instance['Location'].str.split(',', n = 1, expand = True) 
instance['City'] = temp[0].str.strip()
instance['Country'] = temp[1].str.strip()
instance['Check City'] = instance['City'].apply(lambda x: len(gc.get_cities_by_name(x)) + len(gc.get_cities_by_name(x+' City')) )
instance['Check Country'] = instance['Country'].apply(lambda x: 0 if countries.get(x,0)==0 else 1)
instance = instance.query('`Check City`>0 & `Check Country`>0').copy()

temp = instance['Store Potential Sales'].str.split(' ', n = 1, expand = True) 
instance['Store Potential Sales'] = temp[0].str.strip().astype('int64')
instance['Currency'] = temp[1].str.strip()

temp = instance['Store Cost'].str.split(' ', n = 1, expand = True) 
instance['Store Cost'] = temp[0].str.strip().astype('int64')

final = pd.merge(instance,currency,how='inner', on='Currency')[['City', 'Country', 
                'Potential Store Location', 'Store Potential Sales', 'Store Cost','Currency', 'Value in USD']]
final = final.query('`Store Potential Sales`>= `Store Cost`').copy()
final.rename(columns={'Potential Store Location':'Zip Code'}, inplace=True)
final = final.iloc[final.reset_index().groupby('Zip Code')['Store Potential Sales'].idxmax()]
